Add java local https post (#1029)
oktadeveloper/okta-blog: _source/_posts/2022-01-31-local-https-java.adoc

---
layout: blog_post

> title: "HTTPS for Local Java Development"

Reviewer: ```suggestion
title: "Three Ways to Run Your Java Locally with HTTPS"
```
scores way better and describes more

Now access this boring, non-secure application in your browser at`\http://localhost:8080/`.

== HTTPS using an internet proxy

One of the quickest ways to get a local application running with TLS is to use an internet proxy like ngrok.
https://ngrok.com/download[Download and install ngrok] and then start up the service using:

[source,bash]
----
ngrok http 8080
----

This command will expose your local service running port `8080` over the internet on a _generated_ URL such as `\https://abcd-11-22-33-44.ngrok.io`.

If you don't want to install ngrok, you could use https://localhost.run/[localhost.run], which works using SSH port forwarding. Try it out with:

[source,bash]
----
ssh -R 80:localhost:8080 nokey@localhost.run
----

Similar to ngrok, a temporary URL will be generated. For example, `\https://152ccb7ac9369e.lhr.life`.

The goal of these tools is to provide a publicly available URL for your local development, which is perfect if you want to share something via the internet. If you don't want to share, it's best not to expose your development machine to any potential attackers online. The following two options avoid this issue.

== Create a certificate authority with mkcert

> The next two options use a tool called https://github.com/FiloSottile/mkcert[`mkcert`], which sets up a locally trusted certificate authority (CA), installed into the trust stores on your computer. Any certificates issued by this CA will be trusted by the client of your choice (Chrome, Firefox, curl, etc).

Reviewer: ```suggestion
The next two options use a tool called https://github.com/FiloSottile/mkcert[`mkcert`], which sets up a locally trusted certificate authority (CA), installed into the trust stores on your computer. Any certificates issued by this CA will be trusted by the client of your choice (Chrome, Firefox, curl, etc.).
```
fussy punctuation

[source,bash]
----
SPRING_CONFIG_IMPORT='${user.home}/.config/spring-boot/spring-boot-devtools.properties'
----
====

== Bonus: Use a different hostname

The above examples use `localhost`, which works, but you might want to use a different hostname. One easy way to do this is to add a hostname alias to your `/etc/hosts` file:

[source,bash]
----
sudo sh -c 'echo "127.0.0.1 local.example" >> /etc/hosts'
----

In the above examples, when generating certificates with `mkcert`, replace `localhost` with your domain name. Here's an example:

[source,bash]
----
mkcert -key-file devcerts/key.pem -cert-file devcerts/cert.pem local.example
----

WARNING: Use a domain name that you own or that is NOT in the official https://data.iana.org/TLD/tlds-alpha-by-domain.txt[top-level domain list] to ensure there are no conflicting or hijacked DNS entries.

== Potential problems using HTTPS for local development

It's not all sunshine and roses. If you are a developer that works on multiple applications there are a few things to watch out for.

> If you are switching between applications that run on HTTP and HTTPS prototypes, your browser may automatically switch to HTTPS, especially if one of your applications sets the https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security[HTTP Strict-Transport-Security (HTST) headers]. If you run into this type of problem you can clear the https://www.a2hosting.com/kb/getting-started-guide/internet-and-networking/clearing-a-web-browsers-ssl-state[SSL State] and/or https://www.thesslstore.com/blog/clear-hsts-settings-chrome-firefox/[HSTS settings] in your browser.

Reviewer: ```suggestion
Your browser may automatically switch to HTTPS if you are switching between applications that run on HTTP and HTTPS prototypes, especially if one of your applications sets the https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security[HTTP Strict-Transport-Security (HTST) headers]. If you run into this type of problem you can clear the https://www.a2hosting.com/kb/getting-started-guide/internet-and-networking/clearing-a-web-browsers-ssl-state[SSL State] and/or https://www.thesslstore.com/blog/clear-hsts-settings-chrome-firefox/[HSTS settings] in your browser.
```
There's a readability 'rule' engraved in my brain that says you should never start 3 sentences in a row with the same word - so I switched this one around so as not to start with "If..."

[source,bash]
----
ssh -R 80:localhost:8080 nokey@localhost.run
----

Similar to ngrok, a temporary URL will be generated. For example, `\https://152ccb7ac9369e.lhr.life`.

The goal of these tools is to provide a publicly available URL for your local development, which is perfect if you want to share something via the internet. If you don't want to share, it's best not to expose your development machine to any potential attackers online. The following two options avoid this issue.

== Create a certificate authority with mkcert

The next two options use a tool called https://github.com/FiloSottile/mkcert[`mkcert`], which sets up a locally trusted certificate authority (CA), installed into the trust stores on your computer. Any certificates issued by this CA will be trusted by the client of your choice (Chrome, Firefox, curl, etc).

On macOS, you can install `mkcert` with Homebrew; for other operating systems you can find instructions in the https://github.com/FiloSottile/mkcert#installation[`mkcert` docs].

[source,bash]
----
brew install mkcert nss
# nss is only needed if you are using Firefox
----

Create and install the certificate authority:

> ====
> [source,bash]
> ----
> mkcert -install
> ----
> 
> [source,txt]
> ----
> Created a new local CA 💥
> Sudo password:
> The local CA is now installed in the system trust store! ⚡️
> The local CA is now installed in Java's trust store! ☕️
> ----
> ====

Author: ```suggestion
[source,bash]
----
mkcert -install
----

[source,txt]
----
Created a new local CA 💥
Sudo password:
The local CA is now installed in the system trust store! ⚡️
The local CA is now installed in Java's trust store! ☕️
----
```

    proxy_set_header X-Forwarded-Server $host;
    proxy_set_header X-Forwarded-Proto https;
  }

  listen 443 ssl;
  ssl_certificate /etc/nginx/certs/cert.pem;
  ssl_certificate_key /etc/nginx/certs/key.pem;
}
----
<.> Redirect all http requests to https.
<.> Windows and Mac hosts run Docker in a VM, the hostname `host.docker.internal` is used to point back to the "host" OS and the port.
<.> Set the proxy headers, so the downstream app will know what URL the client is using.
====

=== Tie it all together with Docker

Now that you have your Java application, the certificates, and the Nginx configuration ready to go, the only thing left is to start it up!

Use `docker run` to start Nginx with the above files as volumes that configure the service.

====
[source,bash]
----
docker run \
    -v $(pwd)/nginx-default.conf.template:/etc/nginx/templates/default.conf.template \ # <.>
    -v $(pwd)/devcerts:/etc/nginx/certs \ # <.>
    -p 443:443 \
    -p 80:80 \
    --env TARGET_PORT=8080 \ # <.>
    nginx
----
<.> The Nginx configuration file template.
<.> The path to the certificates.
<.> The port where your application is running on `localhost`.
====

> NOTE: You can codify this solution using a https://github.com/bdemers/localhost-tls/blob/main/docker-compose.yml[docker-compose.yml file] if you prefer!

Author: TODO: clean this repo up and fork it in oktadev  -> @bdemers

> NOTE: You can codify this solution using a https://github.com/bdemers/localhost-tls/blob/main/docker-compose.yml[docker-compose.yml file] if you prefer!

Author: ```suggestion
NOTE: You can codify this solution using a https://github.com/oktadev/mkcert-nginx-example/blob/main/docker-compose.yml[docker-compose.yml file] if you prefer!
```
